File: benchmarks-QuantumOptics.jl/qfunc_operator.jl
using QuantumOptics
using BenchmarkTools
include("benchmarkutils.jl")

name = "qfunc_operator"

samples = 5
evals = 20
cutoffs = [10:10:200;]

alpha = 0.7
xvec = collect(linspace(-50, 50, 100))
yvec = collect(linspace(-50, 50, 100))

function f(state, xvec, yvec)
    qfunc(state, xvec, yvec)
end

println("Benchmarking: ", name)
print("Cutoff: ")
results = []
for N in cutoffs
    print(N, " ")
    b = FockBasis(N-1)
    state = coherentstate(b, alpha)
    op = state ⊗ dagger(state)
    t = @belapsed f($op, $xvec, $yvec) samples=samples evals=evals
    push!(results, Dict("N"=>N, "t"=>t))
end
println()

benchmarkutils.save(name, results)

File: benchmarks-QuTiP/qfunc_operator.py
import qutip as qt
import numpy as np
import benchmarkutils

name = "qfunc_operator"

samples = 3
evals = 5
cutoffs = range(10, 101, 10)

def setup(N):
    alpha = 0.7
    xvec = np.linspace(-50, 50, 101)
    yvec = np.linspace(-50, 50, 101)
    state = qt.coherent(N, alpha)
    op = state*state.dag()
    return op, xvec, yvec

def f(state, xvec, yvec):
    return qt.qfunc(state, xvec, yvec, g=2)

print("Benchmarking:", name)
print("Cutoff: ", end="", flush=True)
checks = {}
results = []
for N in cutoffs:
    print(N, "", end="", flush=True)
    op, xvec, yvec = setup(N)
    alpha_check = 0.6 + 0.1j
    checks[N] = f(op, [alpha_check.real], [alpha_check.imag])[0, 0]
    t = benchmarkutils.run_benchmark(f, op, xvec, yvec, samples=samples, evals=evals)
    results.append({"N": N, "t": t})
print()

benchmarkutils.check(name, checks)
benchmarkutils.save(name, results)
